documentation/design/proposals/crdification.adoc

```diff
@@ -0,0 +1,114 @@
+service-broker components. These will have to change to watch and operate on the CRD resources directly
+rather than using the ConfigMap-based implementation. Libraries already exist for that, so the amount of change is small.
+
+The following validations are performed on an `AddressSpace` and will have to be moved to an OpenAPIv3 spec, or (if not expressable in openapi) as a list of status messages on the address space resource:
+
+* Checking that a valid `plan` and `type` is set. `type` can be an OpenAPIv3 enum. `plan` must be checked by address-space-controller
+* Checking that connector references and endpoints are valid, sufficient with OpenAPIv3.
+* Attaching infraUuid label and annotation. Consider deriving the infraUuid based on namespace + name rather than a random UUID. This will prevent the situation where user can remove the annotation and label on the CR.
+
+=== Address
+
+The `Address` resource is used by the address-space-controller (for quota checks), standard-controller and agent. These components will have to be updated to support using CRDs in the same way as for the `AddressSpace` resource.
+
+The following validations are performed on an `Address` resource:
+
+* Checking that a valid `plan` and `type` is set. `type` can be an OpenAPIv3 enum but must also be checked in per-address-space-type controllers. `plan` must be checked by standard-controller/agent
+* Checking that forwarders point to a valid address space connector. Must be checked by standard-controller.
+* Checking that name is prefixed with a valid address space name. Must be checked by standard-controller/agent.
+
+=== MessagingUser
+
+The messaging-user is only used by the API server today, which uses keycloak to persist users.
+
+With CRDs, a new 'messaginguser-controller' component is needed (added to the enmasse-controller-manager for instance). This
+component will sync messaginguser resources with Keycloak.
+
+Having the `MessagingUser` as a CRDs means that we can no longer 'hide' the password for the `password` authentication type. An option of reading passwords from secrets should be provided when creating messaging users.
+
+Validation on the `MessagingUser` can mostly be done by OpenAPIv3 validation, but name prefix checking for address space name must be done by controller.
+
+== Upgrade process
+
+Upgrading EnMasse to a version that supports CRDs require that resources based on ConfigMaps must be converted to CRD-based resources, before the reconiliation loops start. Otherwise, it could end up with infrastructure being deleted. Converting the `MessagingUser` resource requires listing users in Keycloak and creating the corresponding MessagingUser resources. However, `MessagingUser` resources with `password` specified cannot be converted since Keycloak does not allow reading passwords.
+
+The upgrade process will follow these steps:
+
+1. Service admin must remove the apiservice resources - to prevent new resources to be created through the API server
```

Comment:
The manual steps bother me.  I think they will be operational difficult for groups operating a service.  I wonder if there is a way we can automate?

Here's a sketch.  For this to work, responsibility for deployment of address-space-controller needs to be given to the operator.  This avoids a race that would occur if deployed separately (yaml/ansible). 

1. the operator starts-up,  detects that upgrade is required
1. scales down the address-space-controller and admin deployments to zero.
1. remove the apiservices (extra RBAC required), api-server deploymnet/service.
1. awaits the pods to be gone.
1. upgrades the address-space-controller and scales it up.
1. address-space-controller converts addressspace CM -> CRs / address CM -> CR.  Removes CMs
1. adddress-space-controller updates the admin/agent deployment and deployment back up.

Reply:
I’ll have another look over the weekend, but I think it looks good. I was trying to avoid the extra permissions needed to manipulate apiservices, but maybe it shouldn’t be a concern.

Reply:
Nice idea to have the enmasse-operator deploy the address-space-controller, I think that was the missing piece to be able to automate this nicely.